packages/compiler/react.ts

```diff
@@ -133,13 +147,88 @@ const getDynamicsFromJSX = (
     identifier: t.Identifier | null,
     expression: t.Expression | null,
   ) => {
-    // We need to create identifiers for expressions
-    const id = identifier || t.identifier(`__${count}`);
+    const id = identifier || path.scope.generateUidIdentifier('$');
     dynamics.push({ value: expression, id });
-    count++;
     return id;
   };
 
+  const type = openingElement.name;
+  if (t.isJSXIdentifier(type)) {
+    const componentBinding = path.scope.getBinding(type.name);
+    const component = componentBinding?.path.node;
+
+    if (
+      t.isFunctionDeclaration(component) ||
+      t.isVariableDeclarator(component) ||
+      type.name.startsWith(type.name[0]!.toUpperCase())
+    ) {
+      const createRoot = addNamed(path, 'createRoot', 'react-dom/client', {
+        nameHint: 'createRoot$',
+      });
+      const createElement = addNamed(path, 'createElement', 'react', {
+        nameHint: 'createElement$',
+      });
+
+      const objectProperties: t.ObjectProperty[] = [];
+      for (const attribute of openingElement.attributes) {
+        if (t.isJSXAttribute(attribute)) {
+          if (
+            t.isJSXAttribute(attribute) &&
+            t.isJSXExpressionContainer(attribute.value)
```

Comment:
Here `t.isJSXAttribute` is called twice.

Reply:
Fixed